### PR TITLE
Deprecate fetch-params.sh and bundle Sapling parameters into zcashd

### LINKED ISSUE #621: Move zk-SNARK parameter path to shared app data location
In the alpha branch, zksnark parameters are placed into `~/.zcash-params`. But this is cutting corners. These values are a kind of "static application data" except we download them separately from our source repository.  If we had no concern about our code size, these could be simply bundled into our source repo and distributed with binary downloads.

So, they should live on the filesystem in a place for system-wide static application data, such as somewhere under `/usr/share` on linux as per http://www.pathname.com/fhs/pub/fhs-2.3.html#USRSHAREARCHITECTUREINDEPENDENTDATA

However, having giant binaries is cumbersome and we may wish to fetch these parameters in a similar manner to the rest of the blockchain, in which case they should live in a variable location such as somewhere under `/var/lib/`.

Also, of course, find the appropriate paths for each platform.

### LINKED ISSUE #671: Review paths on Windows and macOS for zcash params.
This is the original #629 raised from the dead.

### LINKED ISSUE #1155: P2P distribution of proving parameters
As noted in #826 , right now do not have P2P distribution of the proving parameters, might overload our webserver (or huge S3 bill since they're on S3). Solution: P2P distribution?

### LINKED ISSUE #1312: New zkSNARK parameters file system schema + zcashd & fetch-params.sh support.
- parameter files live in `~/.zcash-params/{proving,verifying}.${SHA256_HEX}`
- each ChainParams specifies which sha256 it expects in zcash. **Edit:** this is an app code change.
- fetch-params specifies the hashes it wants for each of the chainparams (maybe 2 or 3 distinct params?)
- if fetch-params doesn't see the file it wants, it wgets to a tempfile, calculates the sha256 hash, constructs the destination filename and does an mv.
- if the mv fails fetch-params _ignore it_ and continues to the next param file.
- after that, checks that the expected path exists and has the right sha256 checksum. The path may not exist if it fetched a different version from the server (thus the sha256 is different) or another process deleted the file. The checksum would fail if something corrupted the files. In either of these cases, check-params exits with a loud error.
- we host all params files with their full sha256 names and keep old version around.

**Edit 2016-08-31:** I'm moving this item which require zcash app code changes, into a separate followup ticket:
- zcashd always verifies the sha256sum before passing the file on to the deserializer. (Or better would be to verify it as it loads to prevent reading the file twice.)

Note some intentional features of this approach:
- every distinct set of parameters has a unique unambiguous filename, so there's no confusion of which version of parameters lives in a given path.
- if the file downloaded from the webserver doesn't match expectations _it is still saved for diagnosis/forensics_, and the user sees the file they downloaded clearly logged, then an error message pointing out a different sha256 was expected.
- users can freely run different versions of zcashd and run the fetch-params of each different release and there are no conflicts or problems.
- there's no lock file. Concurrent fetch-params processes will both succeed successfully (as they should, since there's no problem there. Well, one problem is duplicate bandwidth.)
- zcashd detects quickly if the correct params are not available instead of trying to deserialize or instantiate the wrong parameters.
- Since filenames all have sha256 hashes, when users post error diagnostics, we can quickly identify which parameter version they are using, and we can notice unknown sha256 sums which would be something we should carefully investigate.

Note: This was originally posted to a PR comment https://github.com/zcash/zcash/pull/1310#issuecomment-242881154 -but that was off topic for that PR, since this is a followup feature change.

### LINKED ISSUE #2607: Port zcash-fetch-params to rust
I'm creating a seperate issue to follow up on the discussion with @str4d from #2597:

>> Since zcash seems to be into rust, is there interest to have those scripts translated? I can see why shell might be preferable, but I could offer porting them if you give me some instructions on how you would like to have it integrated into the repo. (I'd prefer to handle this in a followup pull request though).
>
> Interesting idea. Are you envisaging that we build a binary for downloading parameters instead of using a script? It would certainly be more portable, and it would also be a self-contained chunk of code which would make integration easier. We are still figuring out how exactly we will integrate Rust into Zcash, but a separate binary like this would likely be done in a different way to rewriting or extending internal zcashd code. You should open an issue for discussing this!

This is basically what I'm suggesting. :)

From the top of my head, I can think of the following:
- Add it to librustzcash and export a function that can be called from zcashd with C++
- `cargo init` a small project in a subfolder, compile it during the build and install the binary as `zcash-fetch-params`. If the build system changes to cargo, this can be moved to `src/bin/` and cargo is going to build it as one of the resulting binaries.

### LINKED ISSUE #2695: Faster download for proving keys
I am waiting already almost an hour and still two hours more to go, to get that thing. Seriously, this is a problem. Why is it not being hosted on torrents? I have a 9Mb/s connection. I should have 868Mb in under 5 minutes.

I would be more than happy to contribute to a swarm of torrent seeds to make this happen.

### LINKED ISSUE #3147: On *nix systems, prefer parameters to live in Filesystem Hierarchy Standard location.
**Background:**

Currently the parameters live in `~/.zcash-params`. This allows a normal user to run `fetch-params` without root privileges. However, it means that the parameters are duplicated on multi-user systems, and also they cannot be installed as a package.

The parameters are public, immutable data. According to [the Filesystem Hierarchy Standard](http://www.pathname.com/fhs/pub/fhs-2.3.html) they should live under `/usr/share/`.

**Requirements:**

To close this ticket: change the parameter load logic on *nix systems to check *both* `/usr/share/zcash/params/` as well as `$HOME/.zcash-params`. The first found parameters are loaded. If those are invalid the process should exit with an error (and *not* try to load from the other location, because corrupt parameters are always a serious problem requiring user attention).

**Related tickets:**

- #3145 is a feature request for configurable parameter location. If that is also implemented, then the logic for this ticket needs to look in the two paths above *only if* the parameter location option is absent.
- #1312 applies to the contents of the directory, so it's orthogonal to this ticket (they can either be implemented individually or both implemented).

### LINKED ISSUE #3933: Bundle all ≥2.0 zkSNARK params except the Sprout-on-Groth proving params into the binary
Currently users need to download the zkSNARK parameters with `zcash-fetch-params` prior to using their node the first time. This is an annoying hurdle.

What if the params came with the binary distribution? Well, they are large, so that may be cumbersome. However, the size of the parameters may be dominated by the Sprout proving key which is only necessary for users who need to send _from_ Sprout addresses. New users shouldn't ever need to send from Sprout addresses and old users have probably already fetched the parameters.

So, if all the parameters aside from the Sprout-on-groth proving key were small enough, we could bake them into the binary, and leave `zcash-fetch-params` to retrieve only the Sprout-on-groth spending key. Then the wallet could detect when sending from Sprout if the proving parameters are missing and request to the user to fetch them. Most users would never hit that case, presumably, and so most users would have an improved onboarding experience with one fewer steps and startup delay.

### LINKED ISSUE #5503: Make PARAMS_DIR of fetch-params.sh configurable from CLI
When using fetch-params.sh the download location is **hardcoded** and cannot be changed without modifying the script. Please allow the download location of the parameters to be changed by flag to script, or via env variable. Can't the script simply check first if the env $PARAMS_DIR is set and if not, then use the defaults?

### PR DESCRIPTION
**Pull request by ebfull**
This PR bundles Sapling's parameters and Sprout's verifying key into the `zcashd` binary. Sprout's proving key is large and needs to be loaded from disk at proving time if it's to be used, but since Sprout is deprecated and rarely used, we can treat this as an edge case. (Already, our code does not keep the Sprout proving key around in memory and lazily attempts to load it at proving time.) As such, there is no longer a need for a script to download and store parameter files separately from `zcashd`.

I take the step of deprecating the `fetch-params.sh` file in this PR; we can remove it in a future release.

Fixes #3933. Closes #621. Closes #671. Closes #1155. Closes #1312. Closes #2607. Closes #2695. Closes #3147. Closes #5503.

### CODE REVIEW

File: zcutil/fetch-params.sh [hunk 1]
```diff
@@ -3,289 +3,13 @@
...
-SCRIPT_NAME=$(basename $0)
+cat <<EOF
 
-if [[ -z "${XDG_CACHE_HOME:+x}" ]]; then
-    XDG_CACHE_HOME="${HOME}/.cache"
-fi
+This script is no longer needed for users of zcashd; parameters are
+now bundled with the zcashd binary and do not need to be downloaded
+and stored separately. This script is deprecated and will be
+removed in a future release.
```
Comment (daira) [inline zcutil/fetch-params.sh:12]:
```suggestion
This script is no longer needed for users of zcashd, except to spend funds
from the deprecated Sprout pool. The parameters for other pools are now
bundled with the zcashd binary or generated when it is run, and do not
need to be downloaded or stored separately. This script is deprecated and
will be removed in a future release.

If you need to spend Sprout funds, download the Sprout parameter file from
https://download.z.cash/downloads/sprout-groth16.params and put it at
```

File: zcutil/fetch-params.sh [hunk 2]
```diff
@@ -3,289 +3,13 @@
 export LC_ALL=C
 set -eu
```
Comment (daira) [inline zcutil/fetch-params.sh:5]:
```suggestion

uname_S=$(uname -s 2>/dev/null || echo not)

if [ "$uname_S" = "Darwin" ]; then
    PARAMS_DIR="$HOME/Library/Application Support/ZcashParams"
else
    PARAMS_DIR="$HOME/.zcash-params"
fi

```

File: zcutil/fetch-params.sh [hunk 3]
```diff
@@ -3,289 +3,13 @@
...
-SCRIPT_NAME=$(basename $0)
+cat <<EOF
 
-if [[ -z "${XDG_CACHE_HOME:+x}" ]]; then
-    XDG_CACHE_HOME="${HOME}/.cache"
-fi
+This script is no longer needed for users of zcashd; parameters are
+now bundled with the zcashd binary and do not need to be downloaded
+and stored separately. This script is deprecated and will be
+removed in a future release.
```
Comment (daira) [inline zcutil/fetch-params.sh:11]:
```suggestion
This script is no longer needed for users of zcashd, except to spend funds
from the deprecated Sprout pool. The parameters for other pools are now
bundled with the zcashd binary or generated when it is run, and do not
need to be downloaded or stored separately. This script is deprecated and
will be removed in a future release.

If you need to spend Sprout funds, download the Sprout parameter file from
https://download.z.cash/downloads/sprout-groth16.params and put it at
$PARAMS_DIR/sprout-groth16.params
```

File: Cargo.lock [hunk 4]
```diff
@@ -967,6 +967,7 @@ dependencies = [
  "tracing-core",
  "tracing-subscriber",
  "uint",
+ "wagyu-zcash-parameters",
```
Comment (daira) [inline Cargo.lock:970]: I would like to block this PR on checking that `wagyu-zcash-parameters` results in the same parameters as loading the old parameter files. Perhaps we can check this automatically against the same hashes that were used in `fetch-params.sh`?

File: src/rust/src/rustzcash.rs [hunk 5]
```diff
@@ -136,58 +136,48 @@ fn fixed_scalar_mult(from: &[u8; 32], p_g: &jubjub::SubgroupPoint) -> jubjub::Su
...
-        let (spend_path, output_path, sprout_path) = (
-            Path::new(&spend_path),
-            Path::new(&output_path),
-            sprout_path.as_ref().map(Path::new),
-        );
+        let sprout_path = sprout_path.as_ref().map(Path::new);
+
+        let sprout_vk = {
+            use bellman::groth16::{VerifyingKey, prepare_verifying_key};
+            let sprout_vk_bytes = include_bytes!("sprout-groth16.vk");
```
Comment (daira) [inline src/rust/src/rustzcash.rs:167]: I also need to check that this matches the verifying key from the old parameters.

File: src/rust/src/rustzcash.rs [hunk 6]
```diff
@@ -136,58 +136,48 @@ fn fixed_scalar_mult(from: &[u8; 32], p_g: &jubjub::SubgroupPoint) -> jubjub::Su
...
+        };
 
         // Load params
-        let params = load_parameters(spend_path, output_path, sprout_path);
-        let sapling_spend_params = params.spend_params;
-        let sapling_output_params = params.output_params;
+        let (sapling_spend_params, sapling_output_params) = {
+            let (spend_buf, output_buf) = wagyu_zcash_parameters::load_sapling_parameters();
+            let spend_params = Parameters::<Bls12>::read(&spend_buf[..], false)
+            .expect("couldn't deserialize Sapling spend parameters");
```
Comment (daira) [inline src/rust/src/rustzcash.rs:176]: This is odd formatting.

**Review by daira — CHANGES_REQUESTED** [review]
Concept ACK. This affects consensus code so would need three full ACKs. TODO:
* check that `wagyu-zcash-parameters` gives the same parameters as were previously downloaded;
* check that `sprout-groth16.vk` matches the previous Sprout-on-Groth16 verification key;
* since we no longer check the hash of the downloaded `sprout-groth16.params` in `fetch-params.sh`, we should do so in zcashd when we load it.

**Comment by daira** [comment]
I confirmed that `sprout-groth16.vk` (as added by dfa11e1e3087105438fbb94f6e9e7d7cac5235a2) is the first 1828 bytes of `sprout-groth16.params` as downloaded and checked by `zcutil/fetch-params.sh` before this PR. I also checked that 1828 bytes is the correct length of the verification key prefix as of zcash_proofs 0.11.0, bellman 0.14.0, and bls12_381 0.8.0 (which are the locked versions for zcashd v5.5.0):

* https://github.com/zcash/zcash/blob/v5.5.0/Cargo.lock
  * bellman 0.14.0 is not tagged but https://github.com/zkcrypto/bellman/commit/e137775023a647716793a362ace008e058679b2a is the release PR merge
  * bls12_381 0.8.0 is not tagged but https://github.com/zkcrypto/bls12_381/commit/7de7b9d9c509b9973b35a3241b74bbbea95e700a is the release PR merge
* https://github.com/zcash/librustzcash/blob/zcash_proofs-0.11.0/zcash_proofs/src/lib.rs#L363-L368
* https://github.com/zkcrypto/bellman/blob/e137775023a647716793a362ace008e058679b2a/src/groth16/mod.rs#L159-L220
  * `VerifyingKey<Bls12>::read` reads 3 $\mathbb{G}_1$ and 3 $\mathbb{G}_2$ points, then a 4-byte big-endian count field `ic_len`, then `ic_len` $\mathbb{G}_1$ points
  * `read_g1` reads 96 bytes: https://github.com/zkcrypto/bls12_381/blob/7de7b9d9c509b9973b35a3241b74bbbea95e700a/src/g1.rs#L910-L914
  * `read_g2` for `Bls12` reads 192 bytes: https://github.com/zkcrypto/bls12_381/blob/7de7b9d9c509b9973b35a3241b74bbbea95e700a/src/g2.rs#L1055-L1059
  * `ic_len` in `sprout-groth16.{vk,params}` is `0x0000000a` = 10
  * $3 \times 96 + 3 \times 192 + 4 + 10 \times 96 = 1828$

File: src/rust/src/rustzcash.rs (hunk 5 above)
Comment (ebfull) [reply src/rust/src/rustzcash.rs:167]: I'm resolving this because you've seemingly checked that now.

**Comment by ebfull** [comment]
@daira I think it would be redundant to check the hashes of the parameters now. We should check that they're correct and consistent with the old parameters as part of review, of course, but:

* the sprout-groth16 verification key is being checked into this repository, so it would be redundant to check its hash
* we can review the `wagyu-zcash-parameters` crates and see that they produce the correct parameters, and if satisfied, the `Cargo.lock` file acts as a hash that prevents us from unintentionally changing the parameters. I think in this case it is similarly redundant to check the hash.

I'll add a hash check for the Sprout parameters since those aren't being checked anywhere, but that has no effect on the consensus rules since it's just used for the prover. It would be nice though to mitigate user error.

File: Cargo.lock (hunk 4 above)
Comment (ebfull) [reply Cargo.lock:970]: If it helps, the `wagyu-zcash-parameters` crate also has a test you can invoke to check the parameters. Just do `cargo vet inspect wagyu-zcash-parameters 0.2.0 --mode=local` and copy the parameters to the right place (`mkdir src/archived && cp ~/.zcash-params/sapling-spend.params src/archived/sapling-spend.params && cp ~/.zcash-params/sapling-output.params src/archived/sapling-output.params`) and run `cargo test --release` to check.

Comment (daira) [reply Cargo.lock:970]: I did it a different way: I audited the entire code (excluding tests) of `wagyu-zcash-parameters*` as vendored (i.e. in `depends/<arch>/vendored-sources/wagyu-zcash-parameters*` after a build on this PR branch, which ensures I'm auditing code with the same hashes as in `Cargo.lock`), and observed that:
* it concatenates `wagyu-zcash-parameters-$N/src/sapling-spend-$N.params` for N = {1..5} to get `sapling-spend.params`;
* it reads `wagyu-zcash-parameters-6/src/sapling-output-1.params` to get `sapling-output.params`.

So I did this:
```shell
$ cat wagyu-zcash-parameters-1/src/sapling-spend-1.params wagyu-zcash-parameters-2/src/sapling-spend-2.params wagyu-zcash-parameters-3/src/sapling-spend-3.params wagyu-zcash-parameters-4/src/sapling-spend-4.params wagyu-zcash-parameters-5/src/sapling-spend-5.params >sapling-spend.params
$ diff -b sapling-spend.params ~/.zcash-params/sapling-spend.params 
$ diff -b wagyu-zcash-parameters-6/src/sapling-output-1.params ~/.zcash-params/sapling-output.params
```
No output from `diff -b` means the files are identical.

File: src/rust/src/rustzcash.rs [hunk 7]
```diff
@@ -587,19 +578,34 @@ pub extern "C" fn librustzcash_sprout_prove(
...
 
-    drop(sprout_fs);
+        let hash = blake2b_simd::Params::new()
+            .hash_length(64)
+            .hash(&sprout_params_file);
+
+        // b2sum sprout-groth16.params
+        if hash.as_bytes() != hex::decode("e9b238411bd6c0ec4791e9d04245ec350c9c5744f5610dfcce4365d5ca49dfefd5054e371842b3f88fa1b9d7e8e075249b3ebabd167fa8b0f3161292d36c180a").unwrap().as_slice() {
+            panic!("hash of Sprout groth16 parameters file is incorrect");
+        }
```
Comment (daira) [inline src/rust/src/rustzcash.rs:604]: +1 for checking the hash before calling `Parameters::read`. This Is The Way.

Comment (daira) [reply src/rust/src/rustzcash.rs:604]: I checked (using Python 3.11.2's `hashlib.blake2b` as well as `b2sum`) that this hash matches the Blake2b hash of the `sprout-groth16.params` I'd previously downloaded.

**Review by daira — APPROVED** [review]
utACK e6ff0e34c3613c4c7e909843da7343037cf3f024

**Comment by daira** [comment]
Removed the following unneeded text by @ebfull from the description:
>  I'm open to suggestions:
>
> * should we inform the user automatically that the parameter files are no longer needed and can be deleted?
> * ~should we maintain the `fetch-params.sh` script specifically for the Sprout proving key, if users need it?~ unnecessary
> * ~what language should we use in the `fetch-params.sh` deprecation message and the release notes? I gave it a shot but would like feedback.~ I like the current language after @daira's edits.

File: src/rust/src/rustzcash.rs [hunk 8]
```diff
@@ -587,19 +578,34 @@ pub extern "C" fn librustzcash_sprout_prove(
...
-    let sprout_fs =
-        File::open(unsafe { &SPROUT_GROTH16_PARAMS_PATH }.as_ref().expect(
+    let params = {
+        use std::io::Read;
+
+        // Load parameters from disk
+        let sprout_fs = File::open(unsafe { &SPROUT_GROTH16_PARAMS_PATH }.as_ref().expect(
             "Parameters not loaded: SPROUT_GROTH16_PARAMS_PATH should have been initialized",
         ))
         .expect("couldn't load Sprout groth16 parameters file");
```
Comment (daira) [inline src/rust/src/rustzcash.rs:588]:
```suggestion
        let sprout_path = unsafe { &SPROUT_GROTH16_PARAMS_PATH }.as_ref().expect(
            "Parameters not loaded: SPROUT_GROTH16_PARAMS_PATH should have been initialized",
        );
        let how_to_fix =
            "Please download this file from https://download.z.cash/downloads/sprout-groth16.params\n" +
            "and put it at " + sprout_path;
        let sprout_fs = File::open(sprout_path)
            .expect("Couldn't load Sprout Groth16 parameters file.\n" + how_to_fix);
```

File: src/rust/src/rustzcash.rs [hunk 9]
```diff
@@ -587,19 +578,34 @@ pub extern "C" fn librustzcash_sprout_prove(
...
 
-    let mut sprout_fs = BufReader::with_capacity(1024 * 1024, sprout_fs);
+        let mut sprout_fs = BufReader::with_capacity(1024 * 1024, sprout_fs);
 
-    let params = Parameters::read(&mut sprout_fs, false)
-        .expect("couldn't deserialize Sprout JoinSplit parameters file");
+        let mut sprout_params_file = vec![];
+        sprout_fs
+            .read_to_end(&mut sprout_params_file)
+            .expect("couldn't read Sprout groth16 parameters file");
```
Comment (daira) [inline src/rust/src/rustzcash.rs:595]:
```suggestion
            .expect(&("Couldn't read Sprout Groth16 parameters file.".to_owned() + &how_to_fix));
```

File: src/rust/src/rustzcash.rs [hunk 10]
```diff
@@ -587,19 +578,34 @@ pub extern "C" fn librustzcash_sprout_prove(
...
 
-    drop(sprout_fs);
+        let hash = blake2b_simd::Params::new()
+            .hash_length(64)
+            .hash(&sprout_params_file);
+
+        // b2sum sprout-groth16.params
+        if hash.as_bytes() != hex::decode("e9b238411bd6c0ec4791e9d04245ec350c9c5744f5610dfcce4365d5ca49dfefd5054e371842b3f88fa1b9d7e8e075249b3ebabd167fa8b0f3161292d36c180a").unwrap().as_slice() {
+            panic!("hash of Sprout groth16 parameters file is incorrect");
```
Comment (daira) [inline src/rust/src/rustzcash.rs:603]:
```suggestion
            panic!("Hash of Sprout Groth16 parameters file is incorrect.{}", &how_to_fix);
```

File: src/rust/src/rustzcash.rs [hunk 11]
```diff
@@ -587,19 +578,34 @@ pub extern "C" fn librustzcash_sprout_prove(
...
+        }
+
+        Parameters::read(&sprout_params_file[..], false)
+            .expect("couldn't deserialize Sprout JoinSplit parameters file")
```
Comment (daira) [inline src/rust/src/rustzcash.rs:607]:
```suggestion
        Parameters::read(&sprout_params_file[..], false).expect(
            &("Couldn't deserialize Sprout Groth16 parameters file.".to_owned() + &how_to_fix),
        )
```

Comment (daira) [inline src/rust/src/rustzcash.rs:607]:
```suggestion
            .expect("Couldn't deserialize Sprout Groth16 parameters file")
```
There is no point in using `how_to_fix` here: we already read the file and checked its hash, and so whatever the problem is, it is not that the file needs to be re-downloaded.

File: src/rust/src/rustzcash.rs (hunk 8 above)
Comment (daira) [reply src/rust/src/rustzcash.rs:588]: I tested that this works:
```
$ qa/rpc-tests/remove_sprout_shielding.py
[...]
Tests successful

$ mv ~/.zcash-params/sprout-groth16.params ~/.zcash-params/sprout-groth16.old

$ qa/rpc-tests/remove_sprout_shielding.py
Initializing test directory /tmp/testi2w0ixzo/1050425
waiting for async operation opid-0b385277-0132-4af6-8397-0db8974f392a
thread '<unnamed>' panicked at 'Couldn't load Sprout Groth16 parameters file.
Please download this file from https://download.z.cash/downloads/sprout-groth16.params
and put it at /home/daira/.zcash-params/sprout-groth16.params: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/rust/src/rustzcash.rs:594:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Unexpected exception caught during testing: [Errno 111] Connection refused

$ #download https://download.z.cash/downloads/sprout-groth16.params
$ mv ~/Downloads/sprout-groth16.params ~/.zcash-params/

$ qa/rpc-tests/remove_sprout_shielding.py
[...]
Tests successful
```

**Comment by daira** [comment]
I think the test failures are CI weather.

**Comment by nuttycom** [comment]
This needs a rebase, and potentially fixes given the number of CI failures.

File: src/rust/src/sprout-groth16.vk
Comment (str4d) [inline src/rust/src/sprout-groth16.vk:1]: I verified that this file is the same as the first `1828` bytes of `sprout-groth16.params`:
```
$ head -c 1828 ~/.zcash-params/sprout-groth16.params | b2sum
847777e71ceea9d06ef47ac0c980c2b2a9e65e7617033908f621e067e8c0460242addb6c4bf1e2850c24fe3f9f6e63b9cade4fbf531f5321a1e93c447786c05d  -
$ cat src/rust/src/sprout-groth16.vk | b2sum
847777e71ceea9d06ef47ac0c980c2b2a9e65e7617033908f621e067e8c0460242addb6c4bf1e2850c24fe3f9f6e63b9cade4fbf531f5321a1e93c447786c05d  -
```

**Review by str4d — APPROVED** [review]
ACK cf2ab215be0b93abba41aeefba8af6da6c828922.

**Review by nuttycom — APPROVED** [review]
utACK

**Review by daira — APPROVED** [review]
utACK cf2ab215be0b93abba41aeefba8af6da6c828922

**Pull request merged**